mutation_writer/multishard_writer.cc

```diff
@@ -131,16 +132,25 @@ future<stop_iteration> multishard_writer::handle_mutation_fragment(mutation_frag
     auto f = make_ready_future<>();
     if (mf.is_partition_start()) {
         _consumed_partitions++;
-        if (unsigned shard = shard_for_mf(mf); shard != _current_shard) {
-            _current_shard = shard;
-            if (!bool(_shard_writers[shard])) {
-                f = make_shard_writer(shard);
+        if (auto shards = shard_for_mf(mf); shards != _current_shards) {
+            _current_shards = shards;
+            for (auto shard : shards) {
+                if (!bool(_shard_writers[shard])) {
+                    f = f.then([this, shard] {
+                        return make_shard_writer(shard);
+                    });
+                }
             }
         }
     }
     return f.then([this, mf = std::move(mf)] () mutable {
-        assert(_current_shard != -1u);
-        return _queue_reader_handles[_current_shard]->push(std::move(mf));
+        assert(!_current_shards.empty());
+        if (_current_shards.size() == 1) [[likely]] {
+            return _queue_reader_handles[_current_shards[0]]->push(std::move(mf));
+        }
+        return seastar::parallel_for_each(_current_shards, [this, mf = std::move(mf)] (unsigned shard) {
+            return _queue_reader_handles[shard]->push(mutation_fragment_v2(*_s, mf.permit(), mf));
+        });
```

Comment:
This patch changes the capabilities of multishard_writer to duplicate streams and not just redirect them. Its description should be updated (or created).

Reply:
```patch
index ebb43f3578..80300e9444 100644
--- a/mutation_writer/multishard_writer.hh
+++ b/mutation_writer/multishard_writer.hh
@@ -15,6 +15,21 @@
 
 namespace mutation_writer {
 
+/// Distribute a mutation stream across shards according to the provided sharder.
+/// The function returns a future that is ready when all shards are done consuming.
+///
+/// The provided "consumer" function is called on the destination shard to create a consumer
+/// on that shard. It takes a flat_mutation_reader_v2 that will contain only the mutations that
+/// should be applied on that shard.
+/// The consumer function should return a future that is ready when the shard is done consuming.
+///
+/// Mutations are distributed to shards according to the sharder::shard_for_writes().
+/// During intra-node migration of a tablet shard_for_writes() may return two shards for tokens of that tablet,
+/// in which case that part of the stream may be duplicated to two shards.
+///
+/// The caller is responsible for ensuring that effective_replication_map_ptr used to obtain the sharder
+/// is alive until the returned future is ready. Alternatively, if auto_refreshing_sharder is used,
+/// the topology_guard associated with operation must be used and checked by the consumer.
 future<uint64_t> distribute_reader_and_consume_on_shards(schema_ptr s,
     const dht::sharder& sharder,
     flat_mutation_reader_v2 producer,

```